### DISCUSSION TITLE
How to keep keep Keypadlinc, Fanlinc and HA in sync?

### DISCUSSION THREAD

**Discussion by hlyi** [TD22057/insteon-mqtt#479] [Q&A]
My setup has a 6-button KPL controlling fanlinc ( buton 3-6 of KPL are radio type and are mapped to fan speed high/med/low/off via insteon scenes).  It has been integrated into homeassistant via insteon-mqtt. HA can control the light and fan without any issue. However, the command from HA to Fanlinc doesn't change LEDs on KPL. I updated scenes.yaml file to match all scenes used between PLM/KPL/FanLinc. It didn't help either. How do I update insteon-mqtt configuration files to keep HA/KPL/FanLinc in sync? Thanks!

**Comment by kpfleming** [comment]
When HA activates the fan is it turning on the scene for that selected speed, or is it sending a command directly to the FanLinc? It's probably the latter (that is the default behavior), so the KPL button doesn't have anything to respond to since it doesn't 'snoop' on direct commands to the FanLinc.

**Reply by hlyi** [reply]
HA sends a direct command to FanLinc because MQTT config only provides default settings from the Fanlinc, not scene information. My question was how should I configure insteon-mqtt to make them sync? Is there a way in insteon-mqtt that can send additional command to KPL after receiving a command (from MQTT) to the Fanlinc?

**Comment by kpfleming** [comment]
Not today, no. Ideas for how to build a configuration which could do this are certainly welcome, but note that at the moment insteon-mqtt doesn't even have the ability to send the 'turn LED on' command to a KPL.

**Reply by hlyi** [reply]
Thanks @kpfleming . I will post an idea in "ideas" section.

**Comment by TD22057** [comment]
See the docs here: https://github.com/TD22057/insteon-mqtt/blob/master/docs/mqtt.md#keypadlinc
Insteon-mqtt can control the LED's on the non-load button via direct on/off commands (this is not the same as clicking the button - it just sets the LED state).  It can also simulate the scene command as if you pressed the button via the scene command.  The scene commands are generally what you want - they simulate a button press.

So if you have a keypadlinc button that does what you want (change the fan state), you can have home assistant send a scene command which will trigger that state.  What I don't know is if the keypad will honor the radio settings in that case and turn the other LED's off - we had talked about implementing that at some point but I'm not sure where it's at right now.  You could implement an automation in home assistant to do that using the direct on/off commands to insure the radio lights are kept in sync.

**Reply by hlyi** [reply]
@TD22057 Thanks for the suggestion. Actually I started with HA built-in insteon integration first, and created a automation that does what you said. However, it run into some race-condition, I could improve it. But I moved to insteon-mqtt and hope additional layer of abstraction could handle it robustly. I posted an idea in a different thread. Please let me know what your thoughts. Thanks,

**Comment by TD22057** [comment]
That's my point - it already can I think.  If you send a scene command to the keypadlinc, that triggers the scene which will cause the fanlinc to update and the keypadlinc to update.  That should also emit the state update messages from each device to update HA w/ the correct states.  The simplest way to try this out is to skip HA and just use mosquitto_pub (or something similar) to send some MQTT commands directly and see what happens.  

If you decide to create a switch in HA that sends multiple on/off commands (change fan, change LED's) I'm not sure what you mean by "race condition" - as far as I can see there isn't anything that conflict so there can't be a race.  I don't believe calling on/off on the LED's triggers their scene so the only interaction is calling on/off on the fanlinc (which also wouldn't trigger the scene AFAIK).

**Reply by hlyi** [reply]
When I used HA built-in insteon integration with custom automation, the problem was using KPL: for example, sometime, when I turned on light via KPL, the light turned on, but immediately turned off (by automation). I didn't find a good way to debug it in HA, my theory was that when pressing button KPL, the PLM received multiple state change messages (from scene change, KPL change, fanlinc change), it causes temporary state change that triggered automation, which turned the light off. It didn't happen all the time. That is the race condition I referred.

**Comment by kpfleming** [comment]
Ahh, sorry for the misinformation above... clearly I haven't been keeping up with feature development here :-)
